File: core/transfer/lanescheduling/performance.go
// Package lanescheduling owns content completion estimates and bounded exploration.
// Route permission and transport liveness remain with their existing owners.
package lanescheduling

import "time"

const (
	InitialBlockTime  = 250 * time.Millisecond
	MinimumSampleTime = time.Millisecond
	MeasurementWindow = time.Second
	RelayCostFactor   = 1.1
	sampleWeight      = 0.25
)

// Performance measures authenticated payload over a lane's busy interval, so
// overlapping requests do not each claim the lane's entire throughput.
type Performance struct {
	PendingBytes        uint64
	BytesPerSecond      float64
	HasSuccessfulSample bool
	LastAttempt         time.Time
	busySince           time.Time
	completedBytes      uint64
}

func (p *Performance) Begin(now time.Time, bytes uint64) {
	if p.PendingBytes == 0 {
		p.busySince = now
		p.completedBytes = 0
	}
	p.PendingBytes += bytes
	p.LastAttempt = now
}

func (p *Performance) Complete(now time.Time, bytes uint64, successful bool) {
	if successful && bytes != 0 {
		p.recordAuthenticated(now, bytes)
	}
	p.PendingBytes -= min(p.PendingBytes, bytes)
}

func (p *Performance) recordAuthenticated(now time.Time, bytes uint64) {
	p.completedBytes += bytes
	elapsed := max(now.Sub(p.busySince), MinimumSampleTime)
	if !p.HasSuccessfulSample || elapsed >= MeasurementWindow || bytes >= p.PendingBytes {
		sample := float64(p.completedBytes) / elapsed.Seconds()
		// Congestion should immediately reduce admission. Capacity growth is
		// smoothed so a short burst cannot flood a path with new prefix blocks.
		if p.BytesPerSecond == 0 || sample < p.BytesPerSecond {
			p.BytesPerSecond = sample
		} else {
			p.BytesPerSecond += sampleWeight * (sample - p.BytesPerSecond)
		}
		p.HasSuccessfulSample = true
	}
	if elapsed >= MeasurementWindow {
		p.busySince = now
		p.completedBytes = 0
	}
}

// Superseded is a censored observation, not a path failure. It prevents a lane
// that never finishes before cancellation from retaining an optimistic estimate.
func (p *Performance) Superseded(bytes uint64, elapsed time.Duration) {
	if bytes == 0 || elapsed < MinimumSampleTime {
		return
	}
	upperBound := float64(bytes) / elapsed.Seconds()
	if p.BytesPerSecond == 0 || upperBound < p.BytesPerSecond {
		p.BytesPerSecond = upperBound
	}
}

func (p *Performance) Estimate(bytes uint64) time.Duration {
	bytes = max(bytes, 1)
	seconds := InitialBlockTime.Seconds() * float64(p.PendingBytes+bytes) / float64(bytes)
	if p.BytesPerSecond > 0 {
		seconds = float64(p.PendingBytes+bytes) / p.BytesPerSecond
	}
	// Bound both arithmetic and hedge timers when a path barely makes progress.
	return time.Duration(min(max(seconds, MinimumSampleTime.Seconds()), time.Hour.Seconds()) * float64(time.Second))
}

func Cost(estimate time.Duration, relayed bool) float64 {
	if relayed {
		return float64(estimate) * RelayCostFactor
	}
	return float64(estimate)
}

File: web/src/content/scheduling/performance.ts
export const INITIAL_BLOCK_MILLISECONDS = 250
export const MINIMUM_SAMPLE_MILLISECONDS = 1
export const RELAY_COST_FACTOR = 1.1
const SAMPLE_WEIGHT = 0.25
const MEASUREMENT_WINDOW_MILLISECONDS = 1000
const MAXIMUM_ESTIMATE_MILLISECONDS = 60 * 60 * 1000

/** Busy-interval payload throughput counts overlapping work only once. */
export class LanePerformance {
  pendingBytes = 0
  bytesPerSecond = 0
  hasSuccessfulSample = false
  lastAttempt: number | undefined
  #busySince = 0
  #completedBytes = 0
  #pendingBlocks = 0

  begin(now: number, bytes: number): void {
    if (this.pendingBytes === 0) {
      this.#busySince = now
      this.#completedBytes = 0
    }
    this.#pendingBlocks += 1
    this.pendingBytes += bytes
    this.lastAttempt = now
  }

  complete(now: number, bytes: number, successful: boolean): void {
    if (successful && bytes > 0) {
      this.#completedBytes += bytes
      const elapsed = Math.max(now - this.#busySince, MINIMUM_SAMPLE_MILLISECONDS)
      if (!this.hasSuccessfulSample || elapsed >= MEASUREMENT_WINDOW_MILLISECONDS || bytes >= this.pendingBytes) {
        const sample = this.#completedBytes * 1000 / elapsed
        // Reduce admission immediately on congestion, but smooth short capacity bursts.
        this.bytesPerSecond = this.bytesPerSecond === 0 || sample < this.bytesPerSecond
          ? sample
          : this.bytesPerSecond + SAMPLE_WEIGHT * (sample - this.bytesPerSecond)
        this.hasSuccessfulSample = true
      }
      if (elapsed >= MEASUREMENT_WINDOW_MILLISECONDS) {
        this.#busySince = now
        this.#completedBytes = 0
      }
    }
    this.#pendingBlocks -= 1
    this.pendingBytes -= Math.min(this.pendingBytes, bytes)
  }

  /** A lost race bounds speed without misclassifying caller cancellation as failure. */
  superseded(bytes: number, elapsed: number): void {
    if (bytes <= 0 || elapsed < MINIMUM_SAMPLE_MILLISECONDS) return
    const upperBound = bytes * 1000 / elapsed
    if (this.bytesPerSecond === 0 || upperBound < this.bytesPerSecond) {
      this.bytesPerSecond = upperBound
    }
  }

  estimateQueueMilliseconds(): number {
    // Only complete blocks expose progress. Charge outstanding reservations until
    // they settle; a new request's size must never redefine this existing work.
    const milliseconds = this.bytesPerSecond > 0
      ? this.pendingBytes * 1000 / this.bytesPerSecond
      : this.#pendingBlocks * INITIAL_BLOCK_MILLISECONDS
    return Math.min(MAXIMUM_ESTIMATE_MILLISECONDS, milliseconds)
  }

  estimateCompletionMilliseconds(bytes: number): number {
    const service = this.bytesPerSecond > 0
      ? Math.max(bytes, 1) * 1000 / this.bytesPerSecond
      : INITIAL_BLOCK_MILLISECONDS
    return Math.min(MAXIMUM_ESTIMATE_MILLISECONDS,
      Math.max(MINIMUM_SAMPLE_MILLISECONDS, this.estimateQueueMilliseconds() + service))
  }
}

export function completionCost(milliseconds: number, relayed: boolean): number {
  return milliseconds * (relayed ? RELAY_COST_FACTOR : 1)
}
